flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/InputProcessorUtil.java

```diff
@@ -54,6 +54,10 @@ public static CheckpointedInputGate createCheckpointedInputGate(
 			config, inputGate.getNumberOfInputChannels(), taskName, toNotifyOnCheckpoint);
 		registerCheckpointMetrics(taskIOMetricGroup, barrierHandler);
 
+		if (config.isUnalignedCheckpointsEnabled()) {
+			inputGate.registerBufferReceivedListener(barrierHandler);
+		}
```

Comment:
nit: We could always register the listener, but aligned handler could be just ignoring the calls. It would make more OO design.